### PR TITLE
Fix for PHPUnit 10 #5790

### PR DESCRIPTION
**Pull request by junichi11**
- https://github.com/apache/netbeans/issues/5790
- Add `PhpUnitVersion`
- Don't use `NetBeansSuite.php` with PHPUnit 10 because `suite()` method is invoked no longer and we have to extend `TestCase`
- See: https://github.com/sebastianbergmann/phpunit/commit/36354a9cdc69ecabd20ac505b07541cfdde50aaa

nb-php-gh-5790-phpunit10-project-properties

### CODE REVIEW

**Comment by junichi11** [comment]
The suite class file is checked whether it is a subclass of `TestCase`:
https://github.com/sebastianbergmann/phpunit/blob/da707f3748773fbffd20289ae039e5e667d6b3a6/src/Runner/TestSuiteLoader.php#L42-L80

**Comment by junichi11** [comment]
@tmysik What do you think? Any ideas?

File: php/php.phpunit/src/org/netbeans/modules/php/phpunit/ui/customizer/CustomizerPhpUnit.java
Comment (lbownik) [inline php/php.phpunit/src/org/netbeans/modules/php/phpunit/ui/customizer/CustomizerPhpUnit.java:647]: how about
setSelectedItem(preselected != null ? preselected : PhpUnitVersion.getDefault());

Comment (tmysik) [reply php/php.phpunit/src/org/netbeans/modules/php/phpunit/ui/customizer/CustomizerPhpUnit.java:647]: Not needed, no? Or am I missing something?

**Review by tmysik — APPROVED** [review]

**Comment by tmysik** [comment]
Overall looks good to me. Here are my notes:

- cannot we use the PHPUnit way also for older versions of PHPUnit? Maybe we could avoid the PHPUnit version selection at all?
- cannot we detect what version of PHPUnit is used? I mean, some _cheap_ way, like existence of some file inside PHPUnit directory (not so nice but could work)
- cannot we improve error message so user knows what to do if the running of the tests does not work? Something like "Verify PHPUnit version in the Project Properties dialog"

Thanks!

File: php/php.phpunit/src/org/netbeans/modules/php/phpunit/ui/customizer/CustomizerPhpUnit.java
Comment (lbownik) [reply php/php.phpunit/src/org/netbeans/modules/php/phpunit/ui/customizer/CustomizerPhpUnit.java:647]: it is a syntactic proposal , one line of code instead of five, feel free to ignore if you don't like it.

Comment (tmysik) [reply php/php.phpunit/src/org/netbeans/modules/php/phpunit/ui/customizer/CustomizerPhpUnit.java:647]: I am just checking whether we are preventing some kind of error here, like NPE or similarly.

Comment (junichi11) [reply php/php.phpunit/src/org/netbeans/modules/php/phpunit/ui/customizer/CustomizerPhpUnit.java:647]: Or, I may remove this if I fix this issue without selecting the version.

File: php/php.phpunit/src/org/netbeans/modules/php/phpunit/commands/PhpUnit.java
Comment (tmysik) [inline php/php.phpunit/src/org/netbeans/modules/php/phpunit/commands/PhpUnit.java:410]: Maybe we could add a method into `PhpUnitVersion`, e.g. `usesNbSuite()` and call it here?

File: php/php.phpunit/src/org/netbeans/modules/php/phpunit/commands/PhpUnit.java
Comment (tmysik) [inline php/php.phpunit/src/org/netbeans/modules/php/phpunit/commands/PhpUnit.java:414]: And just to verify - is this the final change? It does not seem to me to be; or am I wrong? Thanks.

**Comment by junichi11** [comment]
> * cannot we use the PHPUnit 10 way also for older versions of PHPUnit? Maybe we could avoid the PHPUnit version selection at all?

Maybe, PHPUnit 9 works the same way (for a single file and directory). However, I'm not sure about the others...

> * cannot we detect what version of PHPUnit is used? I mean, some _cheap_ way, like existence of some file inside PHPUnit directory (not so nice but could work)
> * cannot we improve error message so user knows what to do if the running of the tests does not work? Something like "Verify PHPUnit version in the Project Properties dialog"

I'll try thinking whether we can run unit test without selecting the version :)
Thanks!

File: php/php.phpunit/src/org/netbeans/modules/php/phpunit/commands/PhpUnit.java
Comment (tmysik) [reply php/php.phpunit/src/org/netbeans/modules/php/phpunit/commands/PhpUnit.java:414]: I am asking because it is - at least, I believe it is - possible to select more test files (using CTRL) and run them (so, test them using PHPUnit). I hope I am not wrong...

File: php/php.phpunit/src/org/netbeans/modules/php/phpunit/commands/PhpUnit.java
Comment (junichi11) [reply php/php.phpunit/src/org/netbeans/modules/php/phpunit/commands/PhpUnit.java:410]: Yes, it's good way!

File: php/php.phpunit/src/org/netbeans/modules/php/phpunit/commands/PhpUnit.java
Comment (junichi11) [reply php/php.phpunit/src/org/netbeans/modules/php/phpunit/commands/PhpUnit.java:414]: Yes, you are right. However, I'm not sure the way to run multiple files...
We could do it via `NetBeansSuite.php`, so far, right?

```
php vendor/bin/phpunit --help
PHPUnit 10.1.1 by Sebastian Bergmann and contributors.

Usage:
  phpunit [options] UnitTest.php
  phpunit [options] <directory>
```

Comment (tmysik) [reply php/php.phpunit/src/org/netbeans/modules/php/phpunit/commands/PhpUnit.java:414]: Exactly. This is the reason why we had to use `NetBeansSuite.php`. I am afraid we need to investigate how/whether this is supported in PHPUnit now...

**Comment by junichi11** [comment]
@tmysik We can detect the version via `phpunit --version`. What do you think?

nb-php-fix-phpunit10

**Comment by tmysik** [comment]
@junichi11 The ideal solution, of course 👍

File: php/php.phpunit/src/org/netbeans/modules/php/phpunit/PhpUnitVersion.java
Comment (tmysik) [inline php/php.phpunit/src/org/netbeans/modules/php/phpunit/PhpUnitVersion.java:54]: Some logging in `else` would be useful, I think.

**Comment by tmysik** [comment]
@junichi11 Looks good to me. However, we still need to solve situation when several files should be tested (e.g. user selects more files, or re-run the failed tests).

File: php/php.phpunit/src/org/netbeans/modules/php/phpunit/PhpUnitVersion.java
Comment (junichi11) [reply php/php.phpunit/src/org/netbeans/modules/php/phpunit/PhpUnitVersion.java:54]: Makes sense :) I'll add it. Thanks!

**Comment by junichi11** [comment]
@tmysik 
> However, we still need to solve situation when several files should be tested (e.g. user selects more files, or re-run the failed tests).
Yes. I'll submit that problem as a new issue after this is merged. (I have no ideas although I investigated a bit...)
Thank you for your review!

**Comment by tmysik** [comment]
@junichi11 You are welcome!

File: php/php.phpunit/src/org/netbeans/modules/php/phpunit/PhpUnitVersion.java
Comment (junichi11) [reply php/php.phpunit/src/org/netbeans/modules/php/phpunit/PhpUnitVersion.java:68]: Fixed: https://github.com/apache/netbeans/compare/bcfb8f7cf2e9aecd60f51ef32aa97f0fabc61cb7..6ac8eac427e4e3e315bf68f10faf6a2cf5d99147

Comment (tmysik) [reply php/php.phpunit/src/org/netbeans/modules/php/phpunit/PhpUnitVersion.java:68]: Nice, thanks!

**Comment by junichi11** [comment]
I'll click the "Ready for review" button to merge this after CI passes. (This PR is merged by the release team.)

**Comment by KacerCZ** [comment]
I tried to find way how to pass multiple files to PHPUnit, but there is no easy way.

Possible solution could be inspired by https://github.com/infection/infection.
It manipulates XML configuration for PHPUnit to suit its needs and creates temporary XML file.
This file is then passed to PHPUnit with `-c`/`--configuration` parameter.
See
- https://github.com/infection/infection/pull/1547
- https://github.com/infection/infection/blob/master/src/TestFramework/PhpUnit/Config/XmlConfigurationManipulator.php

**Proposal**

1. NetBeans creates temporary XML file
   - if config file exists (specified in dialog configuration or default file name) then it copies its content
   - relative paths in new file needs to be updated for new location
   - if config file does not exist it creates new one
2. create its own test suite with unique name (https://docs.phpunit.de/en/10.1/configuration.html#the-testsuite-element)
3. add files/directories to the suite
4. run PHPUnit with parameters `--configuration` and `--testsuite` (https://docs.phpunit.de/en/10.1/textui.html) to test selected files

What do you think?

**Comment by tmysik** [comment]
@KacerCZ 

Thanks for looking into it! I was thinking about this solution also for the original solution (instead of `NetBeansSuite.php`), however, it cannot work in cases when the user already has any existing PHPUnit configuration. Or am I missing something?

cc @junichi11

**Comment by tmysik** [comment]
@KacerCZ 

Let me improve my comment - I meant, can we always "extend" properly any existing configuration provided by the user? I am not so sure...

**Comment by KacerCZ** [comment]
Existing configuration file must be copied to temporary file and modified - add test suite and update relative paths.

Configuration file can be configured in NetBeans or PHPUnit is looking for `phpunit.xml`, `phpunit.dist.xml` or `phpunit.xml.dist` in working directory (root of project).

**Comment by tmysik** [comment]
Cannot the existing configuration contain some suite already? I mean, can we _always_ update it? Sorry if it is clear and my questions are unnecessary 🙈

**Comment by KacerCZ** [comment]
@tmysik No need to apologize.
Configuration file can contain one or more suites already defined.
Idea is to add new test suite definition with and run this new suite.

Name of the suite could be something like "NetBeansTempSuite" with MD5 hash created from pathnames passed from Netbeans to prevent name collision with already existing suite.

**Comment by tmysik** [comment]
Ha, if we can _add_ a suite, then it seems to me to be the best way to go. Thanks for clarification!

**Comment by junichi11** [comment]
Thanks for investigating it. I'll submit that problem as a new issue later.

**Comment by junichi11** [comment]
I noticed that only one file is run with also PHPUnit 9 when we select several files.

1. Select two files (Tested files)
2. Run > Test files

Am I wrong something?

**Comment by tmysik** [comment]
@junichi11 

I can see 2 files were run, no?

**Comment by junichi11** [comment]
@tmysik I should have written what I ran about that picture, sorry. I selected the directory to show failed results for 2 files. I'm not sure how to run the selected 2 files actually. So, I've looked into it a bit.

https://github.com/apache/netbeans/blob/e65310a45ade2f08ce756536ce83bcc49237b101/php/php.project/src/org/netbeans/modules/php/project/ui/actions/support/CommandUtils.java#L299-L309

It seems that the first item is returned when we select several files.

https://github.com/apache/netbeans/blob/e65310a45ade2f08ce756536ce83bcc49237b101/php/php.project/src/org/netbeans/modules/php/project/ui/actions/support/ConfigActionTest.java#L172-L179

**Comment by tmysik** [comment]
Uff :) Sorry, it has been a long time... Cannot we call `filesForContextOrSelectedNodes()`? But maybe the current behavior is "good enough" and running tests in a folder is what users do? Anyway, thanks for looking into it.

**Comment by junichi11** [comment]
> Cannot we call `filesForContextOrSelectedNodes()`?

We should be able to do it. Then, we have to fix `getTestRunInfoForFile()` or add `getTestRunInfoForFiles()`.

> But maybe the current behavior is "good enough"

Yes, I also think so :)

> and running tests in a folder is what users do?

I'm not sure... but maybe, usage is the following via CLI, so users didn't report anything about running several test files, I guess.

```
Usage:
  phpunit [options] UnitTest.php
  phpunit [options] <directory>
```

It doesn't seem a big problem for users (I think we need not hurry) because there is no report from users for a long time.
At least, PHPUnit 10 should work with this PR if there are no wrong fixes in my changes :)

**Comment by tmysik** [comment]
I totally agree, not a big problem definitely.

BTW one more use-case is, if I remember it correctly 🙂, to rerun only failed tests. But that cannot work even now so all fine 👍 Let's create a ticket for it and merge this PR if there are no objections.

Thanks!

**Comment by junichi11** [comment]
@tmysik I'll do that later :) This PR will be merged by the release team because I would like to include this in NB18. (We must not do it.) : https://lists.apache.org/thread/s3hnjtmr6t7dkyq1ofn4hdqmd0qh2d11
Thank you!

**Comment by neilcsmith-net** [comment]
> This PR will be merged by the release team because I would like to include this in NB18.

Don't worry, I've been following the conversation! :smile:

I know it's been moved out of draft, but wasn't 100% sure from the ongoing conversation if it is ready to merge.  Please confirm, and I'll merge by tomorrow in time for 18-rc2.

**Comment by junichi11** [comment]
Created the issue: https://github.com/apache/netbeans/issues/5880

**Pull request merged**
